### PR TITLE
GH-120982: Add stack check assertions to generated interpreter code

### PR DESCRIPTION
**Pull request by markshannon**
Skipping news as this doesn't fix anything, it just guards against future bugs.

* Issue: gh-120982

### CODE REVIEW

**Review by Fidget-Spinner — APPROVED** [review]

File: Python/ceval_macros.h [hunk 1]
```diff
@@ -246,6 +246,8 @@ GETITEM(PyObject *v, Py_ssize_t i) {
 #define STACK_SHRINK(n)        BASIC_STACKADJ(-(n))
 #endif
 
+#define  WITHIN_STACK_BOUNDS() \
```
Comment (kumaraditya303) [inline Python/ceval_macros.h]:
```suggestion
#define WITHIN_STACK_BOUNDS() \
```
Two spaces seem redundant

File: Tools/jit/template.c [hunk 2]
```diff
@@ -80,6 +80,9 @@ do {                                                         \
 #undef JUMP_TO_ERROR
 #define JUMP_TO_ERROR() PATCH_JUMP(_JIT_ERROR_TARGET)
 
+#undef WITHIN_STACK_BOUNDS
+#define  WITHIN_STACK_BOUNDS() 1
```
Comment (kumaraditya303) [inline Tools/jit/template.c]: Same

File: Python/optimizer_analysis.c [hunk 3]
```diff
@@ -301,6 +301,11 @@ remove_globals(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer,
 
 
 #define STACK_LEVEL()     ((int)(stack_pointer - ctx->frame->stack))
+#define STACK_SIZE()      ((int)(ctx->frame->stack_len))
+
+#define  WITHIN_STACK_BOUNDS() \
```
Comment (kumaraditya303) [inline Python/optimizer_analysis.c]: Same

File: Tools/jit/template.c [hunk 4]
```diff
@@ -80,6 +80,9 @@ do {                                                         \
 #undef JUMP_TO_ERROR
 #define JUMP_TO_ERROR() PATCH_JUMP(_JIT_ERROR_TARGET)
 
+#undef WITHIN_STACK_BOUNDS
+#define WITHIN_STACK_BOUNDS() 1
```
Comment (brandtbucher) [inline Tools/jit/template.c:84]: The failing CI just didn't like the reference to `entry_frame`. I think this should work?

```suggestion
#define WITHIN_STACK_BOUNDS() (STACK_LEVEL() >= 0 && STACK_LEVEL() <= STACK_SIZE())
```

Comment (markshannon) [reply Tools/jit/template.c:84]: We don't need to check this in JIT builds, so defining it to 1 is fine.

Comment (brandtbucher) [reply Tools/jit/template.c:84]: Why don’t we need to check it in JIT builds?

**Pull request merged**